[torch][segment_reduce] Add support for sum and min reductions

serhaty opened the pull request:
Summary:
This concludes the support for all reductions types initially planned (min, max, mean, sum).

Next Steps:
- Cleanups
       -  update default values when length is 0 and initial not given
       - templatize the code to avoid branching with every item.( and other known improvements)
- more unit tests, verification
- benchmarking

Test Plan: updated unit tests.

Differential Revision: D29268218

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -50,8 +54,12 @@ Tensor _segment_reduce_cpu_kernel(
+                reduction == SegmentReductionType::SUM) {
               initial_value = 0;
+            } else if (reduction == SegmentReductionType::MIN) {
+              initial_value = std::numeric_limits<scalar_t>::max();
```
This should be numeric_limits::infinity (and -infinity for reduction type max), this should work now that you dispatch to floating types only.

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -99,8 +112,13 @@ __global__ static void segment_reduce_forward_kernel(
       initial_value = initial_value + data;
+    } else if (reduction == SegmentReductionType::MIN) {
+      initial_value =
+          at::_isnan(data) ? data : std::min<scalar_t>(initial_value, data);
```
infinity here also

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -172,6 +191,12 @@ __global__ static void segment_reduce_backward_kernel(
       int64_t starting_index = (j * stride_count) + lane_id;
       grad_input_data[starting_index] = grad_val;
     }
+  } else if (reduction == SegmentReductionType::SUM) {
```
it's possible that for multi-d sum, backward can be done more efficiently with index_select, but it requires benchmarks (no changes needed, note for the future)

ngimel commented on test/test_segment_reductions.py:
```diff
@@ -83,21 +85,29 @@ def _test_common(
         check_backward = True
 
-        for reduction in ("max", "mean"):
+        for reduction in reductions:
```
it looks like you are only testing with defined initial_value, would be good to test with undefined also. Unfortunately, some cuda cases will break in unsafe mode, so you need to avoid that.

ngimel commented on test/test_segment_reductions.py:
```diff
@@ -83,21 +85,29 @@ def _test_common(
                 expected_grad = [1.0, 0.5, 0.5, 0.333, 0.333, 0.333]
+            elif reduction == "min":
+                initial_value = 1000  # some high number
+                expected_result = [1, float("nan"), 4, initial_value]
```
instead of creating expected_result as a list, you can create is as a torch.tensor with correct device and dtype, and not convert in the _test_common:
```
expected_result = torch.tensor([1, float("nan"), 4, initial_value], device=device, dtype=dtype)
```

ngimel commented on test/test_segment_reductions.py:
```diff
@@ -181,17 +221,31 @@ def test_multi_d(self, device, dtype):
             if reduction == "max":
+                initial_value = 0
                 expected_result = [
                     np.full((2, 5), initial_value).tolist(),
```
same here, can create expected directly as torch tensor (using `torch.full`, `torch.amax`)

ngimel reviewed (commented):
This looks good, left minor comments. 
At this point, you have to start thinking about proper testing (e.g. testing not only successful cases, but that it throws the expected errors for invalid inputs).

Commit:
[torch][segment_reduce] Add support for sum and min reductions (#60379)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/60379

This concludes the support for all reductions types initially planned (min, max, mean, sum).

Next Steps:
- Cleanups
       -  update default values when length is 0 and initial not given
       - templatize the code to avoid branching with every item.( and other known improvements)
- more unit tests, verification
- benchmarking

Test Plan: updated unit tests.

Differential Revision: D29268218

fbshipit-source-id: e399ae0237083ede670ec998cd55a19eca109e19

serhaty commented:
> This looks good, left minor comments.
> At this point, you have to start thinking about proper testing (e.g. testing not only successful cases, but that it throws the expected errors for invalid inputs).

Sounds good. I also noted this down in PR's summary. Will refactor and add more rigorous testing (will also update some default values) in next PR.

ngimel reviewed (approved):
Looks good, let's see what the tests say.
